staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -424,30 +424,124 @@ func (wc *watchChan) startWatching(watchClosedCh chan struct{}, initialEventsEnd
 	close(watchClosedCh)
 }
 
-// processEvent processes events from etcd watcher and sends results to resultChan.
-func (wc *watchChan) processEvent(wg *sync.WaitGroup) {
-	defer wg.Done()
+// processEvents processes events from etcd watcher and sends results to resultChan.
+func (wc *watchChan) processEvents(wg *sync.WaitGroup) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.ConcurrentWatchObjectDecode) {
+		processing := concurrentOrderedEventProcessing{
+			input:       wc.incomingEventChan,
+			processFunc: wc.transform,
+			output:      wc.resultChan,
 
+			objectType:    wc.watcher.objectType,
+			groupResource: wc.watcher.groupResource,
+		}
+		processing.run(wc.ctx, wg, processEventConcurrency)
+	} else {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case e := <-wc.incomingEventChan:
+					res := wc.transform(e)
+					if res == nil {
+						continue
+					}
+					if len(wc.resultChan) == outgoingBufSize {
+						klog.V(3).InfoS("Fast watcher, slow processing. Probably caused by slow dispatching events to watchers", "outgoingEvents", outgoingBufSize, "objectType", wc.watcher.objectType, "groupResource", wc.watcher.groupResource)
+					}
+					// If user couldn't receive results fast enough, we also block incoming events from watcher.
+					// Because storing events in local will cause more memory usage.
+					// The worst case would be closing the fast watcher.
+					select {
+					case wc.resultChan <- *res:
+					case <-wc.ctx.Done():
+						return
+					}
+				case <-wc.ctx.Done():
+					return
+				}
+			}
+		}()
+	}
+}
+
+type concurrentOrderedEventProcessing struct {
+	input       chan *event
+	processFunc func(*event) *watch.Event
+	output      chan watch.Event
+
+	// Metadata for logging
+	objectType    string
+	groupResource schema.GroupResource
+}
+
+func (p *concurrentOrderedEventProcessing) run(ctx context.Context, wg *sync.WaitGroup, concurrency int) {
+	processingQueue := make(chan chan *watch.Event, concurrency-1)
```

Comment:
nit: make processingQueue a field of `concurrentOrderedEventProcessing`

Reply:
Done